### PR TITLE
example: add no loss lottery example and README

### PR DESCRIPTION
**Pull request by Vvaradinov**
This PR adds a `NoLossLottery.sol` contract that implements a no-loss-lottery using the `Staking` and `Distribution` extension.

### CODE REVIEW

File: examples/no-loss-lottery/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+the `Staking` and `Distribution` precompiled contracts - [NoLossLottery](./contracts/NoLossLottery.sol)
+
+The contract showcases how you can use the contract itself as a manager of funds to ensure
+that users cannot undelegate their funds at any time skewing the lottery results.
+
+**THIS IS NOT A COPY / PASTE CONTRACT** as it has certain limitations and is generally not production ready.
+
+## Implementation
+
+### Deposits
```
Comment (GAtom22) [inline examples/no-loss-lottery/README.md:13]:
```suggestion
### Deposits

```

File: examples/no-loss-lottery/README.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+### Enter Lottery
+The contract allows users to enter the lottery by using the funds they have deposited into the contract
+to delegate to a validator.
+
+### Draw Lottery
+The contract uses the `Distribution` precompile to calculate the rewards of the total pool of funds. Each user
+is assigned a number of tickets based on the amount of tokens they have delegated. Users are shuffled before each draw
+and the winner is selected based on the number of tickets they have and a random number.
+
+### Withdraw Winnings
```
Comment (GAtom22) [inline examples/no-loss-lottery/README.md:28]:
```suggestion
### Withdraw Winnings

```

File: examples/no-loss-lottery/README.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+delegation rewards have been distributed
+
+### No Multiple Validators
+The contract does not handle users delegating to multiple validators. This is for simplicity’s sake as this would require
+keeping track of which users have delegated to which validators and would require a more complex implementation.
+
+### No Partial Exit Lottery
+The contract does not allow multiple undelegations with smaller amounts from the validator. 
+This is for simplicity’s sake as this would require keeping track of which users have undelegated how much 
+and would require a more complex implementation.
```
Comment (GAtom22) [inline examples/no-loss-lottery/README.md:62]:
```suggestion
and would require a more complex implementation.

```

File: examples/no-loss-lottery/contracts/NoLossLottery.sol [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        // Transfer the requested amount of Evmos to the sender.
+        payable(msg.sender).transfer(_amount);
+        delete unbondingDelegations[msg.sender];
+        emit Withdraw(msg.sender, _amount);
+    }
+
+    // Withdraw winnings from contract
+    function withdrawWinnings() public {
+        require(winnings[msg.sender] > 0, "You have no winnings yet");
+        payable(msg.sender).transfer(winnings[msg.sender]);
```
Comment (GAtom22) [inline examples/no-loss-lottery/contracts/NoLossLottery.sol:78]: ditto

File: examples/no-loss-lottery/contracts/NoLossLottery.sol [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    // Exits the lottery returning the funds back to deposits after the unbonding period
+    function exitLottery(string memory _validatorAddr, uint256 _amount) public {
+        require(_amount <= delegations[msg.sender], "This address does not hold a delegation amount with the lottery");
+        require(_amount == delegations[msg.sender], "You must exit the entire delegation amount");
+        int64 completionTime = STAKING_CONTRACT.undelegate(address(this), _validatorAddr, _amount);
+        delegations[msg.sender] -= _amount;
+        unbondingDelegations[msg.sender] = UnbondingRequest(completionTime, _amount);
+        totalStaked -= _amount;
+        deposits[msg.sender] += _amount;
+        totalTickets -= _amount / 1e18;
```
Comment (GAtom22) [inline examples/no-loss-lottery/contracts/NoLossLottery.sol:136]: nit: ditto, use `SafeMath`

**Review by GAtom22 — APPROVED** [review]
LGTM! Great work @Vvaradinov !! Left a few nit comments

File: examples/no-loss-lottery/README.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
+# No Loss Lottery
+
+This is an example of how to create a smart contract for a no-loss-lottery that utilizes
+the `Staking` and `Distribution` precompiled contracts - [NoLossLottery](./contracts/NoLossLottery.sol)
```
Comment (danburck) [inline examples/no-loss-lottery/README.md:4]: Add context on how a No Loss Lottery works

File: examples/no-loss-lottery/contracts/NoLossLottery.sol [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    // Enters the lottery by delegating to a validator
+    function enterLottery(string memory _validatorAddr, uint256 _amount) public {
+        require(_amount % 1 ether == 0, "Amount must be a whole number of Evmos");
+        require(_amount <= deposits[msg.sender], "This address does not hold a deposit amount with the lottery");
+        _approveRequiredMsgs(_amount);
+        STAKING_CONTRACT.delegate(address(this), _validatorAddr, _amount);
+        delegations[msg.sender] += _amount;
+        deposits[msg.sender] -= _amount;
+        totalStaked += _amount;
+        totalTickets += _amount / 1e18;
```
[the remark this answers is not in the document]
Comment (Vvaradinov) [reply examples/no-loss-lottery/contracts/NoLossLottery.sol:118]: I agree but I think it's a bit overkill for the sake of examples, will definitely add this to the `Limitations` readme section.

**Pull request merged**